source/extensions/filters/network/kafka/kafka_request_parser.h

```diff
@@ -0,0 +1,197 @@
+ * @see http://kafka.apache.org/protocol.html#protocol_messages
+ */
+class RequestHeaderParser : public Parser {
+public:
+  // default constructor
+  RequestHeaderParser(const RequestParserResolver& parser_resolver, RequestContextSharedPtr context)
+      : RequestHeaderParser{parser_resolver, context,
+                            std::make_unique<RequestHeaderDeserializer>()} {};
+
+  // visible for testing
+  RequestHeaderParser(const RequestParserResolver& parser_resolver, RequestContextSharedPtr context,
+                      RequestHeaderDeserializerPtr deserializer)
+      : parser_resolver_{parser_resolver}, context_{context}, deserializer_{
+                                                                  std::move(deserializer)} {};
+
+  /**
+   * Uses data provided to compute request header
+   * @return Parser instance responsible for processing rest of the message
+   */
+  ParseResponse parse(const char*& buffer, uint64_t& remaining) override;
+
+  const RequestContextSharedPtr contextForTest() const { return context_; }
+
+private:
+  const RequestParserResolver& parser_resolver_;
+  const RequestContextSharedPtr context_;
+  RequestHeaderDeserializerPtr deserializer_;
+};
+
+/**
+ * Sentinel parser that is responsible for consuming message bytes for messages that had unsupported
+ * api_key & api_version. It does not attempt to capture any data, just throws it away until end of
+ * message
+ */
+class SentinelParser : public Parser {
+public:
+  SentinelParser(RequestContextSharedPtr context) : context_{context} {};
+
+  /**
+   * Returns UnknownRequest
+   */
+  ParseResponse parse(const char*& buffer, uint64_t& remaining) override;
+
+  const RequestContextSharedPtr contextForTest() const { return context_; }
+
+private:
+  const RequestContextSharedPtr context_;
+};
+
+/**
+ * Request parser uses a single deserializer to construct a request object
+ * This parser is responsible for consuming request-specific data (e.g. topic names) and always
+ * returns a parsed message
+ * @param RequestType request class
+ * @param DeserializerType deserializer type corresponding to request class (should be subclass of
+ * Deserializer<RequestType>)
+ */
+template <typename RequestType, typename DeserializerType> class RequestParser : public Parser {
+public:
+  /**
+   * Create a parser with given context
+   * @param context parse context containing request header
+   */
+  RequestParser(RequestContextSharedPtr context) : context_{context} {};
+
+  /**
+   * Consume enough data to fill in deserializer and receive the parsed request
+   * Fill in request's header with data stored in context
+   */
+  ParseResponse parse(const char*& buffer, uint64_t& remaining) override {
+    const uint64_t orig_remaining = remaining;
+    try {
+      context_->remaining_request_size_ -= deserializer.feed(buffer, remaining);
+    } catch (const EnvoyException&) {
+      // treat the whole request as invalid, throw away the rest of the data
```

Comment:
Question: Does this happen normally? If not, should we avoid trying to be smart and skipping a request vs. just letting this type of exception propagate fully out and we can kill the connection or just stop trying to parse like we do with Mongo?

Reply:
No, it does not (it means that client is sending garbage or deserialization is buggy).
Will propagate the exception - the filter will know best what to do with it (metrics filter can just disable sniffing; thick kafka-facade should stop iteration I think).